From: Herbert Xu <herbert@gondor.apana.org.au>
Date: Fri, 22 Jan 2016 23:17:58 +0800
Subject: crypto: hash - Add helpers to zero stack request/descriptor
References: bnc#974525,FATE#320741
Patch-Mainline: v4.6-rc2
Git-commit: e67ffe0af4d450e95e5f9bb7d081500e30777857

As the size of an ahash_request or shash_desc is variable, it's
awkward to zero them explicitly.  This patch adds helpers to do
that which should be used when they are created on the stack.

Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
Acked-by: Hannes Reinecke <hare@suse.de>
(cherry picked from commit 4853f0cbd3218feb89a3fac61f3232380bfcf22f)
---
 include/crypto/hash.h | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/include/crypto/hash.h b/include/crypto/hash.h
index 6361892..aef8192 100644
--- a/include/crypto/hash.h
+++ b/include/crypto/hash.h
@@ -14,6 +14,7 @@
 #define _CRYPTO_HASH_H
 
 #include <linux/crypto.h>
+#include <linux/string.h>
 
 struct crypto_ahash;
 
@@ -550,6 +551,12 @@ static inline void ahash_request_free(struct ahash_request *req)
 	kzfree(req);
 }
 
+static inline void ahash_request_zero(struct ahash_request *req)
+{
+	memzero_explicit(req, sizeof(*req) +
+			      crypto_ahash_reqsize(crypto_ahash_reqtfm(req)));
+}
+
 static inline struct ahash_request *ahash_request_cast(
 	struct crypto_async_request *req)
 {
@@ -872,4 +879,10 @@ int crypto_shash_final(struct shash_desc *desc, u8 *out);
 int crypto_shash_finup(struct shash_desc *desc, const u8 *data,
 		       unsigned int len, u8 *out);
 
+static inline void shash_desc_zero(struct shash_desc *desc)
+{
+	memzero_explicit(desc,
+			 sizeof(*desc) + crypto_shash_descsize(desc->tfm));
+}
+
 #endif	/* _CRYPTO_HASH_H */
-- 
1.8.5.6

